Extract huc8 geojsons reliably

vlulla opened the pull request:
To generate all the huc8 geojsons (around 2266) we need to ensure a couple of things:

1. All the `nhdflowline` features are present in the `nhdplusv2` table.
2. All the `wbdhu8` boundaries are present in the `nhdplushhr` table.

The sql queries had to be modified because the original sql queries were generating a ton of warnings. These log warnings used up a lot of space on the database docker container and raised the "No space left on the device" error.

The huc8 geojson code generation was modified slightly to only run for geojsons that are not already present in the output dir. Even though docker stats lists that docker is using all the cpus assigned to docker it takes quite a bit of time to generate these geojsons. For instance, it took almost an hour to generate 75 geojsons on my laptop (actually docker with 6 cpus and 12 gb memory). Coincidentally, the battery went from 100% charged to 48% charged in 1.5 hours!

To recreate all the geojsons just delete all the geojsons with `rm *.json`. However, if you are wary of doing that, just do `mv huc8-extracts huc8-extracts-old` and the script will recreate the folder. By the way, this is a good idea...especially if you wish to compare if the geojsons generated are different in any possible way.

Since the geojsons created by `save_huc8_extracts`, during the initial run, had more information than we needed we just transform those geojsons (`transform_geojsons`) into a format that we need for our experiment.

The generated geojsons can be found in s3://azavea-noaa-hydro-data/noaa/huc8-extracts/transformed/

Closes #48

vlulla commented:
@lewfish OOPS, I forgot to link the various issues and understandings that we have gathered. Sorry! I'll update the description with appropriate links in the next few days.

rajadain commented:
What is the location / bucket where the extracts have been stored?

vlulla commented:
Still generating (running on my laptop...). I'm planning on uploading it to s3://azavea-noaa-hydro-data/noaa/huc8-extracts/

lewfish commented:
> The huc8 geojson code generation was modified slightly to only run for geojsons that are not already present in the output dir. Even though docker stats lists that docker is using all the cpus assigned to docker it takes quite a bit of time to generate these geojsons. For instance, it took almost an hour to generate 75 geojsons on my laptop (actually docker with 6 cpus and 12 gb memory). Coincidentally, the battery went from 100% charged to 48% charged in 1.5 hours!

You might be able to parallelize things using Dask to use all the cores.

lewfish commented:
> > The huc8 geojson code generation was modified slightly to only run for geojsons that are not already present in the output dir. Even though docker stats lists that docker is using all the cpus assigned to docker it takes quite a bit of time to generate these geojsons. For instance, it took almost an hour to generate 75 geojsons on my laptop (actually docker with 6 cpus and 12 gb memory). Coincidentally, the battery went from 100% charged to 48% charged in 1.5 hours!
> 
> You might be able to parallelize things using Dask to use all the cores. Although things might be bottlenecked by the database.

Ah, I see you're already using the map functionality in Dask.

lewfish reviewed (changes requested):
There's some duplicated code between the functionality for HUC8 vs. HUC12. I would refactor to get rid of this. Also, as a general rule, you should not commit commented out code. If we need to bring back something that was deleted we can get it from the git history.

vlulla commented:
> There's some duplicated code between the functionality for HUC8 vs. HUC12. I would refactor to get rid of this. Also, as a general rule, you should not commit commented out code. If we need to bring back something that was deleted we can get it from the git history.

Ok, will remove the commented lines. You were right about dask being bottlenecked by database. If geojson generation is going to be a recurring thing we can rearrange the tables so that they are in the same database (but different schemas) and then run the query within that database. That might help with not having to create intermediate shapely object ([en/de]coding issues), and doing fewer database connections.

Once I make your recommended changes, after the geojson finishes (hopefully tonight), should I just create a new commit and push to this branch? Or should I do `git commit -amend`?

lewfish commented:
> Once I make your recommended changes, after the geojson finishes (hopefully tonight), should I just create a new commit and push to this branch? Or should I do `git commit -amend`?

Usually it's good to put these fixes in response to a PR review into separate commits with a title that starts with "fix: ". That way it's easier to review. Then, before merging, you can do an interactive rebase to squash those fix commits into real commits that will get merged.

vlulla commented:
> Usually it's good to put these fixes in response to a PR review into separate commits with titles that starts with "fix: ". That way it's easier to review. Then, before merging, you can do an interactive rebase to squash those fix commits into real commits that will get merged.

Ok, sounds good! Will take a few attempts to get the hang of this workflow. Thanks!

vlulla commented:
> Usually it's good to put these fixes in response to a PR review into separate commits with titles that starts with "fix: ". That way it's easier to review. Then, before merging, you can do an interactive rebase to squash those fix commits into real commits that will get merged.

Removed the commented blocks! Added minor modifications that might be of help. What do I do next?

lewfish commented:
> Removed the commented blocks! Added minor modifications that might be of help. What do I do next?

If you think the requirements in the associated issue have been fulfilled, and you have addressed the requested changes in the review, then you can signal that the PR is ready for another review by saying "Ready for review again" or something.

vlulla reviewed (commented):
I have made your recommended changes.

Closes #48

vlulla reviewed (commented):
Made the recommended changes.

Closes #48

vlulla commented:
> * You forgot the part about removing duplicated code.
> * There is still some commented out code.
> * There is some markdown that refers to HUC 12 that should be changed to HUC 8.
> * There is a mix of single and double quotes. It's good to be consistent and at Azavea we typically use single quotes in Python.
> * You should include any instructions or scripting that was used for loading the databases in case someone else needs to run this again.
> * The `Closes #48` should go in the PR description (ie. the topmost comment block on the PR page) so that it is easily accessible.
> * In Jupyter notebooks there shouldn't be a `main` function that you define in one cell, and then call in another. You should simply put the content of the main function directly into a cell and then execute that cell.
> * I checked the output, and it still contains the stream geometries. As discussed in the last meeting, you should reprocess these files so that they only have the HUC geometry and the list of COMIDs. You can leave the existing files you have since they might be useful and took a long time to create, but I would just add another cell in the notebook that writes a new set of files.

Hopefully, made all the recommended changes. Uploaded these transformed jsons to s3://azavea-noaa-hydro-data/noaa/huc8-extracts/transformed/ . Is this what you were expecting?

Finally, if you want I can change the original `save_huc8_extracts` to incorporate the lessons I've learned and recreate the geojsons in the final format.

lewfish reviewed (changes requested):
There are still some unaddressed points of feedback from the previous review. If something doesn't make sense, please ask for clarification.

* There is still some commented out code.
* There is a mix of single and double quotes. It's good to be consistent and at Azavea we typically use single quotes in Python.
* The Closes #48 should go in the PR description (ie. the topmost comment block on the PR page) so that it is easily accessible.
* In Jupyter notebooks there shouldn't be a main function that you define in one cell, and then call in another. You should simply put the content of the main function directly into a cell and then execute that cell.
* You should include any instructions or scripting that was used for loading the databases in case someone else needs to run this again.

Some new things based on the changes that were made:

* The implementation of `transform_geojson` could be simplified. The outline of the HUC was already available in the original GeoJSON file and the database didn't need to be used. This also would have avoided having to use Pandas and doing the concat with inner join. Since this is a one-off script we probably won't use again you don't need to change this, but it's something to keep in mind for the future.
* You repeated the same comment twice -- "## neither fiona nor geopandas support writing lists to geojson"
* The last cell that does a sort of test of the output was a good thing to include.
* For variable and function names in Python, there is a conventions we generally follow that I also see used in most modern Python projects. The convention is to use an underscore as a separator between words rather than using camel case or having no separators. Some of the code in `transform_geojsons` does not follow these conventions so I would update this. For example, `joinedData` should be `joined_data` and `outdir` should be `out_dir`. (Note: some older Python libraries including the standard library and `numpy` deviate from these. If you were to work on these libraries, then you would want to follow the conventions used in those libraries.)
* Another naming convention is to avoid single letter variable names. These are not very readable and are difficult to do a find and replace on. Some exceptions are variable that are just used within a list comprehension. Also, if it's a numerical algorithm and there is some standard notation using letters like `i` and `j` I think that's ok to use.

I'd like to find a better workflow for reviewing Jupyter notebooks as it's not supported well in Github. I'm looking into https://www.drivendata.co/blog/nbautoexport-jupyter-code-review/

lewfish commented:
> Hopefully, made all the recommended changes. Uploaded these transformed jsons to s3://azavea-noaa-hydro-data/noaa/huc8-extracts/transformed/ . Is this what you were expecting?

Yes, this works.

> 
> Finally, if you want I can change the original `save_huc8_extracts` to incorporate the lessons I've learned and recreate the geojsons in the final format.

That's ok. If we run this again in the future we should change this, but at this point I don't think it's necessary.

vlulla commented:
> There are still some unaddressed points of feedback from the previous review. If something doesn't make sense, please ask for clarification.
> 
> * There is still some commented out code.
> * There is a mix of single and double quotes. It's good to be consistent and at Azavea we typically use single quotes in Python.
> * The Closes #48 should go in the PR description (ie. the topmost comment block on the PR page) so that it is easily accessible.
> * In Jupyter notebooks there shouldn't be a main function that you define in one cell, and then call in another. You should simply put the content of the main function directly into a cell and then execute that cell.
> * You should include any instructions or scripting that was used for loading the databases in case someone else needs to run this again.
> 
> Some new things based on the changes that were made:
> 
> * The implementation of `transform_geojson` could be simplified. The outline of the HUC was already available in the original GeoJSON file and the database didn't need to be used. This also would have avoided having to use Pandas and doing the concat with inner join. Since this is a one-off script we probably won't use again you don't need to change this, but it's something to keep in mind for the future.
> * You repeated the same comment twice -- "## neither fiona nor geopandas support writing lists to geojson"
> * The last cell that does a sort of test of the output was a good thing to include.
> * For variable and function names in Python, there is a conventions we generally follow that I also see used in most modern Python projects. The convention is to use an underscore as a separator between words rather than using camel case or having no separators. Some of the code in `transform_geojsons` does not follow these conventions so I would update this. For example, `joinedData` should be `joined_data` and `outdir` should be `out_dir`. (Note: some older Python libraries including the standard library and `numpy` deviate from these. If you were to work on these libraries, then you would want to follow the conventions used in those libraries.)
> * Another naming convention is to avoid single letter variable names. These are not very readable and are difficult to do a find and replace on. Some exceptions are variable that are just used within a list comprehension. Also, if it's a numerical algorithm and there is some standard notation using letters like `i` and `j` I think that's ok to use.
> 
> I'd like to find a better workflow for reviewing Jupyter notebooks as it's not supported well in Github. I'm looking into https://www.drivendata.co/blog/nbautoexport-jupyter-code-review/

Here are all the changes I made:
- simplified `transform_geojsons` based on your recommendation
- added notes/instructions, with links for appropriate pages, to document what all went into this PR
- functions called in the same cell that they are defined
- modified the top conversation block to contain the closes issue link
- removed unneeded imports
- no more single letter variable names, hopefully
- no more duplicate comments
- no more quoting inconsistencies, hopefully

Please let me know if I've still missed some things!

lewfish reviewed (changes requested):
We're getting closer, but you're still missing two things:

* The quote inconsistency is still there. Perhaps you could use the "Find" functionality in your text editor to search for `'` and `"` to check that this has been done. 
* In Jupyter notebooks there shouldn't be a main function that you define in one cell, and then call in another. You should simply put the content of the main function directly into a cell and then execute that cell.

vlulla commented:
> We're getting closer, but you're still missing two things:
> 
> * The quote inconsistency is still there. Perhaps you could use the "Find" functionality in your text editor to search for `'` and `"` to check that this has been done.
> * In Jupyter notebooks there shouldn't be a main function that you define in one cell, and then call in another. You should simply put the content of the main function directly into a cell and then execute that cell.

Fixed both of these. Please let me know if I've still missed something.

lewfish reviewed (approved):
Looks good now. I would squash all these commits into the first one using an interactive rebase. Once that's done you can merge.
